### ISSUE TITLE
doesn't return the output

### ISSUE THREAD

**Issue by MarioRuiz** [net-ssh/net-ssh#337]
On net-ssh 3.0.2 and previous @output got the output of the command, in this case 'ls', but now on 3.1.0 and 4.0.0.. returns nothing
```ruby
`  @output=""
  command="ls"
  Net::SSH.start(@host, @user, password: @password, port: @port) do |my|
    my.open_channel do |ch|
        ch.exec(command) do |ch,success|
          raise "could not execute command" unless success  
          ch.on_data do |c, data|
            @output=@output + data.to_s()
          end
        end  
    end
  end
  puts @output
```

**Comment by mfazekas** [comment]
There is a wait missing:
```
`  @output=""
  command="ls"
  Net::SSH.start(@host, @user, password: @password, port: @port) do |my|
    my.open_channel do |ch|
        ch.exec(command) do |ch,success|
          raise "could not execute command" unless success  
          ch.on_data do |c, data|
            @output=@output + data.to_s()
          end
        end  
    end
   ch.wait
  end
  puts @output
  
```
see https://github.com/net-ssh/net-ssh/blob/master/lib/net/ssh.rb#L98
https://github.com/net-ssh/net-ssh/blob/master/lib/net/ssh/connection/session.rb#L296

**Comment by MarioRuiz** [comment]
weird that it was not necessary before... thanks!

**Comment by mfazekas** [comment]
Can you try a command that take longer? Like ```command="sleep 5 ; ls"``` I'm sure that reproduces the issue on old version as well.
So the wait was neccesary in old version you just got lucky there IMO.

**Comment by MarioRuiz** [comment]
I tried... and it changes anything... with "sleep 5;ls" on 3.0.2 returns the folder content and not in 3.1.0 or 4.0.0

**Comment by mfazekas** [comment]
So it looks cause is https://github.com/net-ssh/net-ssh/pull/319/files 
in 3.0.1 before what happens is this:
when you get to end of block passed to start it'll close the session which tries to close all the connections:
https://github.com/net-ssh/net-ssh/blob/master/lib/net/ssh/connection/session.rb#L116
in 3.0.1 there was a bug that caused channels that haven't yet opened to ignore close event.
In your case event loop not ran after open_channel so it's such channel, so while close called on it it's ignored so channel opens and communication done.

This should not work on 3.0.2 either:
```ruby
@output=""
  command="sleep 10 ; ls"
  Net::SSH.start(@host, @user, password: @password, port: @port) do |my|
    ch = my.open_channel do |ch|
        ch.exec(command) do |ch,success|
          raise "could not execute command" unless success  
          ch.on_data do |c, data|
            @output=@output + data.to_s()
          end
        end
    end
    count = 0
    my.loop(0.5) { count += 1 ; count < 10  }
    puts "should be opened by now"
  end
  puts @output
```

so this was really a bug in 3.0.2 that caused that to work, and it was fixed in 3.1. In 4.0 i'm open to consider a change Net::SSH.start so it waits for all the opened channels to finish, bit i need to think through it.

**Comment by MarioRuiz** [comment]
I see, thanks a lot for the answer, I was quite curious
Regarding 4.0.... well.. in my opinion it is better to have the freedom to wait or not... maybe in some cases we want just to launch the process without waiting for them to finish
